### PR TITLE
Allow specifying the hyphenation cost

### PR DESCRIPTION
**Pull request by mattfbacon**
This PR adds a `hyphenation-cost` parameter to `text` which allows tuning the hyphenation algorithm.

Feel free to bikeshed the name.

### CODE REVIEW

File: crates/typst-library/src/text/mod.rs [hunk 1]
```diff
@@ -364,6 +364,17 @@ pub struct TextElem {
     #[resolve]
     pub hyphenate: Hyphenate,
 
+    /// The "cost" of hyphenation when laying out text.
+    /// A higher cost means the layout engine will hyphenate less often.
+    ///
+    /// When justifying text, a higher hyphenation cost can result in more variation
+    /// in justification spacing in order to avoid hyphenation.
+    ///
+    /// The default cost is an acceptable balance, but some may find that it hyphenates too eagerly.
```
Comment (andreasKroepelin) [inline crates/typst-library/src/text/mod.rs:373]: Maybe mention that the default is `0.5` here.

Comment (mattfbacon) [reply crates/typst-library/src/text/mod.rs:373]: I figured since the docs show the default it's not necessary to repeat it

Comment (andreasKroepelin) [reply crates/typst-library/src/text/mod.rs:373]: Oh never mind then!

**Comment by mattfbacon** [comment]
Any update on this? Anything I need to do before it's merged?

**Comment by laurmaedje** [comment]
I was/am just very busy with web app issues, I'll get to it soon.

File: crates/typst-library/src/text/mod.rs [hunk 2]
```diff
@@ -364,6 +364,17 @@ pub struct TextElem {
...
+    /// A cost of `10.0` may work better for such contexts.
+    #[default(0.5)]
```
Comment (PgBiel) [inline crates/typst-library/src/text/mod.rs:375]: Is it possible to deduplicate this value with the `const DEFAULT_HYPH_COST` from the other file?

Comment (mattfbacon) [reply crates/typst-library/src/text/mod.rs:375]: Yeah, I was confused about that. `shared_get` returns an Option despite this having a default value. I'm not very familiar with the whole system.

Comment (laurmaedje) [reply crates/typst-library/src/text/mod.rs:375]: shared_get returns `None` if the value is not uniform for the whole paragraph

**Comment by laurmaedje** [comment]
I think have some knob for tweaking hyphenation is great, but exposing the raw cost might not yield the ideal user experience. It is a very opaque metric. Could we perhaps expose this as a ratio between `0%` (few (no?) hyphens) and `100%` (many hyphens).

**Comment by mattfbacon** [comment]
Ok.. would that be as a ratio of MAX_COST? Like 100% = 0 cost, 0% = max?

And if we add configuration for runt prevention, along the same lines, should it be a percentage, or just a boolean?

File: crates/typst-library/src/text/mod.rs (hunk 2 above)
Comment (mattfbacon) [reply crates/typst-library/src/text/mod.rs:375]: So what is the solution there? Is it correct to default to the default cost when shared_get returns None?

**Comment by mattfbacon** [comment]
I'm also wondering about the general shape of these options in the face of new knobs for text shaping, like my runt prevention PR. Should this go into a dictionary named something like `text-shaping` which can group these options?

**Comment by mattfbacon** [comment]
Sorry, I just realized I used the wrong remote branch for this PR so I'm gonna have to close this one and open another with the correct branch.

**Pull request closed without merging**